gwos/contracts/state-validator/src/verifications/submit_block.rs

```diff
@@ -690,58 +732,155 @@ fn check_block_withdrawals(block: &L2BlockReader) -> Result<(), Error> {
+        {
+            debug!(
+                "[check_block_timestamp] block.timestamp is out of available range, block.timestamp: {}, backbone range: [{}-{}, {}+{}]",
+                block_timestamp,
+                backbone, BACKBONE_BIAS,
+                backbone, BACKBONE_BIAS
+            );
+            return Err(Error::InvalidBlock);
+        }
     }
 
-    let rollup_input_timestamp = match rollup_input_since.extract_lock_value() {
-        Some(LockValue::Timestamp(time)) => time,
-        _ => return Err(Error::InvalidSince),
-    };
-    debug!(
-        "[check block timestamp] input since timestamp {}",
-        rollup_input_timestamp
-    );
+    Ok(())
+}
 
-    let tip_block_timestamp = prev_global_state.tip_block_timestamp().unpack();
-    if prev_version > 0 && tip_block_timestamp >= rollup_input_timestamp {
-        debug!("[check block timestamp] input since is smaller than tip block timestamp");
-        return Err(Error::InvalidSince);
+fn check_global_state_tip_block_timestamp(
+    prev_global_state: &GlobalState,
+    post_global_state: &GlobalState,
+) -> Result<(), Error> {
+    let prev_block_timestamp: u64 = prev_global_state.tip_block_timestamp().unpack();
+    let post_block_timestamp: u64 = post_global_state.tip_block_timestamp().unpack();
+    if !(prev_block_timestamp < post_block_timestamp) {
+        debug!(
+            "[check_global_state_tip_block_timestamp] prev_block_timestamp: {}, post_block_timestamp: {}",
+            prev_block_timestamp,
+            post_block_timestamp
+        );
+        return Err(Error::InvalidPostGlobalState);
     }
+    Ok(())
+}
 
-    let post_block_timestamp: u64 = post_global_state.tip_block_timestamp().unpack();
-    if block_timestamp != post_block_timestamp
-        || block_timestamp > rollup_input_timestamp
-        || (prev_version != 0 && block_timestamp <= tip_block_timestamp)
+fn check_global_state_last_finalized_timepoint(
+    rollup_config: &RollupConfig,
+    context: &BlockContext,
+    prev_global_state: &GlobalState,
+    post_global_state: &GlobalState,
+) -> Result<(), Error> {
+    // skip
+    // if context.post_version == 0 { }
+
+    if prev_global_state.last_finalized_block_number().unpack() > post_global_state.last_finalized_block_number().unpack()
     {
         debug!(
-            "[check block timestamp] invalid block timestamp {}, post block timestamp {}, rollup input timestamp {} tip block timestamp {}",
-            block_timestamp,
-            post_block_timestamp,
-            rollup_input_timestamp,
-            tip_block_timestamp
+            "prev_global_state.last_finalized_block_number > post_global_state.last_finalized_block_number"
         );
-        return Err(Error::InvalidBlock);
+        return Err(Error::InvalidPostGlobalState);
+    }
+
+    let last_finalized_timepoint =
+        Timepoint::from_full_value(post_global_state.last_finalized_block_number().unpack());
+
+    if context.post_version == 1 {
```

Comment:
define a function to check feature support:

`if !use_timestamp_as_timepoint()`

Reply:
The new commit 3519b6383b113ee308b4bfe0f39c09d9c72eebe2 add a zero-sized type `Fork` as fork features collection. PTAL.

Reply:
Another new commit 9487d16cf70d933547ae0b0961c8826be22f1f82 do it.